### ISSUE TITLE
get_messages() either segfaults or returns an error with timeout=0

### ISSUE THREAD

**Issue by pokemane** [intrepidcs/python_ics#69]
First, apologies if this isn't the right repo to bring this up in; I'm trying to use pylessard/python-udsoncan, pylessard/python-can-isotp, and hardbyte/python-can to work with UDS, and python-ics as the actual hardware layer.

python-ics by itself works fine, and adding python-can seems fine too. However, when I went to add python-can-isotp, I started getting segfaults when it was trying to receive messages.

I saw that its `CanStack.rx_canbus()` function [was passing `timeout=0`](https://github.com/pylessard/python-can-isotp/blob/master/isotp/protocol.py#L848) to [python-can](https://github.com/hardbyte/python-can/blob/master/can/bus.py#L83), which was passing that on [to python-ics.](https://github.com/hardbyte/python-can/blob/master/can/bus.py#L83) That last [function (`_process_msg_queue()`)](https://github.com/hardbyte/python-can/blob/29a235bc6e51002b243126b778bd9bc318652be9/can/interfaces/ics_neovi/neovi_bus.py#L225) is where `ics.get_messages()` is called, and where the timeout of `0` finally gets to python-ics.

What I found was that changing the [`self.bus.recv(0)`](https://github.com/pylessard/python-can-isotp/blob/master/isotp/protocol.py#L848) to `self.bus.recv(0.000001)` (or any other tiny nonzero number) fixed this and I no longer get any segfaults or SpyMessages with errors.

The docstring for `ics.get_messages()` suggests that it's ok with *any* float, but it's evidently not the case, at least in this situation. Is this a documentation problem or a problem with the function itself?

Again, I'm not sure if it's specifically because I was stacking python-can and python-can-isotp on top, so I apologize if this isn't the right repo. Let me know if I should open this in one of those two repos instead!

**Comment by drebbe-intrepid** [comment]
@pokemane Can you provide a script to reproduce this behavior and provide details on your environment (operating system, python version, 32bit or 64bit, icsneo40.dll version, etc).

**Comment by pokemane** [comment]
We were actually just looking at this, good timing!

It looks like it might be that [this `count` variable](https://github.com/intrepidcs/python_ics/blob/master/src/methods.cpp#L956) is never actually initialized if the timeout is `0`. It never enters the two `for` loops, and the first one is what gives that counter a value, so maybe when it reaches [here](https://github.com/intrepidcs/python_ics/blob/master/src/methods.cpp#L990) it segfaults?

Also here's a quick script:

```py
import ics
import can
import isotp
import logging
import time
from can.interfaces.ics_neovi.neovi_bus import NeoViBus
from can import Message

enable_use_server = True

hs1_bus = NeoViBus(channel=1, bitrate=500000)

def my_error_handler(error):
   logging.warning('IsoTp error happened : %s - %s' % (error.__class__.__name__, str(error)))

addr = isotp.Address(isotp.AddressingMode.Normal_11bits, txid=0x001, rxid=0x002)
stack = isotp.CanStack(hs1_bus, address=addr, error_handler=my_error_handler, params={'stmin':10, 'can_fd':False, 'tx_padding':0xAA})

stack.send(b'12345678')

while stack.transmitting():
   print(stack.sleep_time())
   stack.process()
   time.sleep(stack.sleep_time())

time.sleep(0.1)

hs1_bus.shutdown()
```

`pip freeze` output:
```
$ python -m pip freeze
aenum==2.2.3
astroid==2.3.3
can-isotp==1.5
colorama==0.4.3
isort==4.3.21
lazy-object-proxy==1.4.3
mccabe==0.6.1
pylint==2.4.4
python-can==3.3.2
python-ics==4.3
six==1.13.0
udsoncan==1.9
windows-curses==2.1.0
wrapt==1.11.2
```

**Comment by pokemane** [comment]
That script is only "half" of a two-script setup, the other half is the "threaded message receive" example from the iso-tp project. I know it's trying to send a multiframe message and it'll never get the flow control frame, but it only segfaults when trying to "process" multiframe messages regardless of whether or not there's something on the other end to give it a flow control response back.

**Comment by drebbe-intrepid** [comment]
@pokemane Can you also provide what Intrepid hardware you are using? I'll try to look at this as soon as possible but with the end of the year and holiday coming up it might take me a while.

There is nothing obvious that I can think of off the top of my head why it would be crashing.

**Comment by pokemane** [comment]
@drebbe-intrepid No problem, I appreciate the time!

I'm using a neoVI FIRE 2. I haven't tried with a neoVI FIRE. I've tried (and reproduced) this on HS1 and HS2, but haven't tried any of the other CAN buses at this time.

**Comment by drebbe-intrepid** [comment]
@pokemane sorry for the delay but I believe the issue has been fixed. It will be in 4.4 once I get around to building it.

In the mean time if you have visual studio 2015+ install you can build straight from git with the following command:

```pip install git+https://github.com/intrepidcs/python_ics@v900```

**Comment by drebbe-intrepid** [comment]
@pokemane I've build and released version 4.4. I'm going to close out this issue. Please re-open if you find this didn't fix your issue.
